main.go

```diff
@@ -80,212 +76,235 @@ func main() {
-					}
+				if err := handleServiceChangedUpdate(client, tagConfig, lock, instances, update); err != nil {
+					glog.Errorf(errorTemplate, update.UpdateType, tagName, err)
 				}
-
-				lock.Unlock()
-			default:
-				continue
 			}
 		case <-done:
-			glog.Warningf("Received termination signal for service with tag %s", tag)
+			glog.Warningf("Received termination signal for service with tag %q", tagName)
 			wg.Done()
 			return
 		}
 	}
 }
 
-// NOTE: Maximum length of resource name is 63 and name must start and end with letter or digit
-func makeName(prefix string, name string) string {
-	n := strings.Join([]string{prefix, name}, "-")
-	if len(n) > maximumNameLength {
-		n = strings.TrimLeft(n[:maximumNameLength], "-")
+func handleServiceNewUpdate(
+	urlMapName string,
+	client cloud.Cloud,
+	tagConfig *tagConfiguration,
+	lock *sync.RWMutex,
+) error {
+	lock.Lock()
+	defer lock.Unlock()
+
+	tagName := tagConfig.Name
+
+	if tagConfig.Mode == REUSE_MODE {
+		glog.Infof("Service with tag %q will use externally managed resources", tagName)
+	} else {
+		negName := tagConfig.NetworkEndpointGroup.Name
+		glog.Infof("Creating network endpoint group %q...", negName)
+		if err := client.CreateNetworkEndpointGroup(negName); err != nil {
+			return fmt.Errorf("Can't create network endpoint group %q: %+v", negName, err)
+		}
+		glog.Infof("Created network endpoint group %q", negName)
+
+		hcName := tagConfig.HealthCheck.Name
+		glog.Infof("Creating health check %q...", hcName)
+		if err := client.CreateHealthCheck(hcName, tagConfig.HealthCheck.Path); err != nil {
+			return fmt.Errorf("Can't create health check %q: %+v", hcName, err)
+		}
+		glog.Infof("Created health check %q", hcName)
+
+		bsName := tagConfig.BackendService.Name
+		glog.Infof("Creating backend service %q...", bsName)
+		if err := client.CreateBackendService(bsName, negName, hcName, tagConfig.BackendService.Affinity, tagConfig.BackendService.CDN); err != nil {
+			return fmt.Errorf("Can't create backend service %q: %+v", bsName, err)
+		}
+		glog.Infof("Created backend service %q", bsName)
+
+		glog.Infof("Updating URL map %q...", urlMapName)
+		if err := client.UpdateURLMap(urlMapName, bsName, tagConfig.BackendService.Host, tagConfig.BackendService.Path); err != nil {
+			return fmt.Errorf("Can't update URL map %q: %+v", urlMapName, err)
+		}
+		glog.Infof("Updated URL map %q", urlMapName)
 	}
-	return n
+
+	glog.Infof("Watching service with tag %q...", tagName)
+
+	return nil
+}
+
+func handleServiceDeletedUpdate(
+	client cloud.Cloud,
+	tagConfig *tagConfiguration,
+	lock *sync.RWMutex,
+	instances map[string]*registry.ServiceInstance,
+) error {
+	lock.Lock()
+	defer lock.Unlock()
+
+	var toRemove []gce.NetworkEndpoint
+
+	for k, instance := range instances {
+		toRemove = append(toRemove, gce.NetworkEndpoint{
+			Instance: normalizeInstanceName(instance.Host),
+			IP:       instance.Address,
+			Port:     instance.Port,
+		})
+		delete(instances, k)
```

Comment:
You are modifying a _variable_ sent into a method, this is very indirect and while allowed in Go/C (mainly for the functionality that has to be fast and the things that have to do with system programming), genuinely frowned upon. Please redo this with immutability in mind.